ARCHITECTURE.md

```diff
@@ -0,0 +1,133 @@
+
+#### Declaring order
+
+Something important to realize is that the order in which the dependencies are declared (for a given container) matters. Let's have a look at some code to make things clearer.
+
+```
+container.register<IOidcHandler>("browser:oidcHandler", {
+  useClass: AggregateOidcHandler,
+});
+container.register<IOidcHandler>("browser:oidcHandlers", {
+  useClass: RefreshTokenOidcHandler,
+});
+container.register<IOidcHandler>("browser:oidcHandlers", {
+  useClass: AuthorizationCodeWithPkceOidcHandler,
+});
+container.register<IOidcHandler>("browser:oidcHandlers", {
+  useClass: ClientCredentialsOidcHandler,
+});
+```
+
+Here, `AggregateOidcHandler` is the handler aggregator (as defined in the Handler Pattern section), and `RefreshTokenOidcHandler`, `AuthorizationCodeWithPkceOidcHandler` and `ClientCredentialsOidcHandler` are its underlying handlers. When receiving a request, `AggregateOidcHandler` will first call to its instance of `RefreshTokenOidcHandler` to check if it can handle it. If so, the instance of `RefreshTokenOidcHandler` will handle the request, and the instances of `AuthorizationCodeWithPkceOidcHandler` or `ClientCredentialsOidcHandler` will not be called. This means that it is important to declare the dependencies from the most specialized to the most generic, because if a fallback handler that can handle all requests is declared first, the other more specialized handlers will not be called.
+
+#### Mocks and tests
+
+Dependency injection makes the codebase more flexible, because it is only at runtime that each component will be presented with the dependencies it declared, which means it is easier to add a dependency to a component without changing the whole codebase.
+
+However, mocking dependency injection in test code wouldn't bring any value. Instead, the object we want to test can be constructed with mocked dependencies provided to its constructor. For instance, a class such as
+
+```
+@injectable()
+export default class RefreshTokenOidcHandler implements IOidcHandler {
+  constructor(
+    @inject("node:tokenRefresher") private tokenRefresher: ITokenRefresher,
+    @inject("node:storageUtility") private storageUtility: IStorageUtility
+  ) {}
+  // ...
+```
+
+can be tested as follows: 
+```
+const refreshTokenOidcHandler = new RefreshTokenOidcHandler(
+  someMockedTokenRefresher,
+  someMockedStorageUtility
+);
+```
+
+## Mapping OIDC flows to the code
+
+### Auth code flow
+
+![Module dependencies](./documentation/diagrams/auth_code_flow.svg)
+
+- Discovery: `packages/*/src/login/oidc/IssuerConfigFetcher.ts`
+- Registration: `packages/*/src/login/oidc/ClientRegistrar.ts`
+- OIDC handler: `packages/*/src/login/oidc/oidcHandlers/AuthorizationCodeWithPkceOidcHandler.ts`
+- Handle incoming redirect: `packages/*/src/login/oidc/redirectHandler/*Handler.ts`
```

Comment:
Might be good to mention here why you can't list the specific Handler here. (Because I don't know :P)

Reply:
7b69d5233e25714ca2f18500cf2cdded6f198ea2 is a bit more specific about the handlers that may get called, but really it depends on their `canHandle` conditions. Typically, the list will grow if we implement error handling as I suggest in the dedicated ticket, by adding a dedicated handler that `canHandle` incoming redirects where the redirect IRI contains an `error` query param.

Reply:
Ah, that sounds like a good approach to error handling. How about just mentioning something like "the specific Handler depends on which Handler's `canHandle()` method first returns `true`"?

Reply:
78e9ddd6ab6e555a1f559a7807465b4dd7e12cb1